lightning-background-processor/src/lib.rs

```diff
@@ -272,22 +288,42 @@ impl BackgroundProcessor {
 				// pruning their network graph. We run once 60 seconds after startup before
 				// continuing our normal cadence.
 				if last_prune_call.elapsed().as_secs() > if have_pruned { NETWORK_PRUNE_TIMER } else { FIRST_NETWORK_PRUNE_TIMER } {
-					if let Some(ref handler) = net_graph_msg_handler {
-						log_trace!(logger, "Pruning network graph of stale entries");
-						handler.network_graph().remove_stale_channels();
-						if let Err(e) = persister.persist_graph(handler.network_graph()) {
+					// The network graph must not be pruned while rapid sync completion is pending
+					log_trace!(logger, "Assessing prunability of network graph");
+					let graph_to_prune = match rapid_gossip_sync.as_ref() {
+						Some(rapid_sync) => {
+							if rapid_sync.is_initial_sync_complete() {
+								Some(rapid_sync.network_graph())
+							} else {
+								None
+							}
+						},
+						None => net_graph_msg_handler.as_ref().map_or(None, |handler| Some(handler.network_graph()))
```

Comment:
FWIW, the interface is prone to misuse since the user could potentially pass in both syncs with two different `NetworkGraph`s, but only one would be persisted. Not sure if we really care but it would be a reason to prefer an enum over two `Option`s.

Reply:
True, but with an enum, you cannot use both a rapid sync and a regular message handler. I can make an extra check verifying that the network graph references are the same somehow? Is there a way to do that that will work with Arc?

Reply:
> True, but with an enum, you cannot use both a rapid sync and a regular message handler.

How useful is this for a mobile? Wouldn't the app only be open infrequently? Just trying to understand the use case of having both.

> I can make an extra check verifying that the network graph references are the same somehow? Is there a way to do that that will work with Arc?

Probably some pointer comparison though not sure if we want to do so.

Reply:
> How useful is this for a mobile? Wouldn't the app only be open infrequently? Just trying to understand the use case of having both.

That depends on whether mobile implementations will choose to obtain gossip data in between rapid syncs. If some apps envision being open for up to a minute or two, it's a relatively cheap thing to do in the background, and allows to minimize trust after the initial fast forward. But we're all just guessing here, I can survey some folks who'll be using it.

Reply:
Just occurred to me that we currently use `NetGraphMsgHandler` to handle the `network_update` field of `Event::PaymentPathFailed`. We'll have to do something similar in the rapid-sync-only case.